File: ros2_ws_p/src/action_package/action_package/count_up_to_client.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus
from robot_interfaces.action import CountUpTo

class CountUpToClient(Node):
    def __init__(self):
        super().__init__("count_up_to")

        self.count_up_to_client_ = ActionClient(
            self, CountUpTo, "/count_up_to")
        
    def send_goal(self, target_number, delay):
        self.count_up_to_client_.wait_for_server()
        goal = CountUpTo.Goal()
        goal.target_number = target_number
        goal.delay = delay
        self.count_up_to_client_.send_goal_async(
            goal, feedback_callback = self.goal_feedback_callback).add_done_callback(
                self.goal_response_callback)
    
    def goal_feedback_callback(self, feedback_msg):
        number = feedback_msg.feedback.current_number
        self.get_logger().info("Current number: " + str(number))
        
    def goal_response_callback(self, future):
        self.goal_handle_: ClientGoalHandle = future.result()
        if self.goal_handle_.accepted:
            self.get_logger().info("Goal got accepted")
            self.goal_handle_.get_result_async().add_done_callback(
                self.goal_result_callback)
        else:
            self.get_logger().info("Goal got rejected")

    def goal_result_callback(self, future):
        status = future.result().status
        result = future.result().result
        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info("Success")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error("Aborted")
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warm("Cancelled")

        self.get_logger().info("Result: " + str(result.reached_number))

def main(args = None):
    rclpy.init(args = args)
    node = CountUpToClient()
    node.send_goal(10, 0.5)
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros2_ws_p/src/action_package_cpp/src/count_up_to_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "robot_interfaces/action/count_up_to.hpp"
#include <functional>

using CountUpTo = robot_interfaces::action::CountUpTo;
using CountUpToGoalHandle = rclcpp_action::ClientGoalHandle<CountUpTo>;
using namespace std::placeholders;

class CountUpToClient : public rclcpp::Node
{
    public:
        CountUpToClient() : Node("count_up_to_client")
        {
            count_up_to_client_ = rclcpp_action::create_client<CountUpTo>(
                this,
                "/count_up_to");

            RCLCPP_INFO(this->get_logger(), "[CountUpToClient] Node started.");
        }

        void send_goal(int target_number, double delay)
        {
            count_up_to_client_->wait_for_action_server();
            auto goal = CountUpTo::Goal();
            goal.target_number = target_number;
            goal.delay = delay;
            
            auto options = rclcpp_action::Client<CountUpTo>::SendGoalOptions();
            options.goal_response_callback = std::bind(
                &CountUpToClient::goalResponseCallback, this, _1);
            options.result_callback = std::bind(
                &CountUpToClient::goalResultCallback, this, _1);
            options.feedback_callback = std::bind(
                &CountUpToClient::goalFeedbackCallback, this, _1, _2);

            count_up_to_client_->async_send_goal(goal, options);
        }
    private:
        rclcpp_action::Client<CountUpTo>::SharedPtr count_up_to_client_;

        void goalResponseCallback(const CountUpToGoalHandle::SharedPtr goal_handle)
        {
            if(!goal_handle)  
                RCLCPP_ERROR(this->get_logger(), "Goal was rejected by the server.");   
        }

        void cancelGoal()
        {
            RCLCPP_INFO(this->get_logger(), "Send a cancel goal request");
            count_up_to_client_->async_cancel_all_goals();
        }

        void goalResultCallback(const CountUpToGoalHandle::WrappedResult &result)
        {
            if(result.code == rclcpp_action::ResultCode::SUCCEEDED)
                RCLCPP_INFO(this->get_logger(), "Success");
            else if(result.code == rclcpp_action::ResultCode::ABORTED)
                RCLCPP_ERROR(this->get_logger(), "Aborted");
            else if(result.code == rclcpp_action::ResultCode::CANCELED)
                RCLCPP_WARN(this->get_logger(), "Canceled");
        }

        void goalFeedbackCallback(
            const CountUpToGoalHandle::SharedPtr goal_handle,
            const std::shared_ptr<const CountUpTo::Feedback> feedback)
        {
            (void)goal_handle;
            int number = feedback->current_number ;
            RCLCPP_INFO(this->get_logger(), "Got feedback: %d", number);

            if(number >= 2)
                cancelGoal();
        }
};
int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUpToClient>();
    node->send_goal(10, 1.0);
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}